### ISSUE TITLE
Tests with injected file system errors.

### ISSUE THREAD

**Issue by krasi-georgiev** [prometheus/tsdb#579]
One case it to catch bugs like:
https://github.com/prometheus/tsdb/pull/575
https://github.com/prometheus/tsdb/issues/565 - A test for this woule look like. Inject an error at some point for these funcs and check that no tmp files is left behind.
An example test:
https://github.com/prometheus/tsdb/pull/576

This article is a good start.
https://medium.com/@siddontang/use-fuse-to-inject-failure-to-i-o-deb5f2e7800a

**Comment by geekodour** [comment]
would like to work on this as it's not a high priority one.

**Comment by pborzenkov** [comment]
@krasi-georgiev Alternative approach would be to use a filesystem abstraction interface (something like https://github.com/spf13/afero) and instantiate it with specially crafted fault-injecting implementation for tests.

**Comment by krasi-georgiev** [comment]
@pborzenkov yes I looked at it , but didn't see anything in the interface to trigger different faults. It looks like its main purpose is to NOT use a real file system and avoid the cleanup after each test.

@pborzenkov can you point to some examples how can afero be used for our use case?

**Comment by pborzenkov** [comment]
@krasi-georgiev I meant something like this:
  - switch all TSDB code that accesses FS to use afero interface
  - in normal conditions, instantiate the interface with real FS implementation (taken from afero)
  - in test, instantiate it with some custom (yet to be written) implementation that allows to inject faults (via callback, random or any other method).

I haven't seen such implementation in the wild, just wanted to say that it might be more useful as it may provide much more granular control of where and when to inject faults.

**Comment by krasi-georgiev** [comment]
yes that is a good suggestion, but I would like to see if we can do this without any  additional intrerfaces. Using afero in the actual tsdb code would add using interfaces and in some hot paths this will have some negative performance impact because of  the interface dereferencing. I have noticed that Fabian removed some interfaces in the past to improve some performance. I think the WAL was an interface and now is passed as a simple struct.

**Comment by pborzenkov** [comment]
Yeah, in this case it's probably better to avoid interfaces. Another implementation of granular fault injections might be something like this:
  - insert static faults into points of interest
  - make them noop by default (via build tags)
  - in tests, build with faults enabled and control which ones to enable

E.g. something like this:

`
func OpenMmapFile(path string) (*MmapFile, error) {
    if err := faults.IsFailed("OpenMmapFile"); err != nil {
        return nil, err
    }

    <normal implementation>
}
`

**Comment by krasi-georgiev** [comment]
yep that is another good idea. Just not sure if adding code just for the sake of better testing is a good pattern.
